src/subscription_manager/syspurposelib.py

```diff
@@ -216,5 +237,105 @@ def read_syspurpose():
             syspurpose = json.load(open(USER_SYSPURPOSE))
         except (os.error, ValueError):
             # In the event this file could not be read treat it as empty
+            if raise_on_error:
+                raise
             syspurpose = {}
     return syspurpose
+
+
+def write_syspurpose(values):
+    """
+    Write the syspurpose to the file system.
+    :param values:
+    :return:
+    """
+    if SyspurposeStore is not None:
+        sp = SyspurposeStore(USER_SYSPURPOSE)
+        sp.contents = values
+        sp.write()
+    else:
+        # Simple backup in case the syspurpose tooling is not installed.
+        try:
+            json.dump(values, open(USER_SYSPURPOSE), ensure_ascii=True, indent=2)
+        except OSError:
+            log.warning('Could not write syspurpose to %s' % USER_SYSPURPOSE)
+            return False
+    return True
+
+
+class SyspurposeSyncActionInvoker(certlib.BaseActionInvoker):
+    """
+    Used by rhsmcertd to sync the syspurpose values locally with those from the Server.
+    """
+
+    def _do_update(self):
+        action = SyspurposeSyncActionCommand()
+        return action.perform()
+
+
+class SyspurposeSyncActionReport(certlib.ActionReport):
+    name = "Syspurpose Sync"
+
+
+class SyspurposeSyncActionCommand(object):
+    """
+    Sync the system purpose values, by performing a three-way merge between:
+      - The last known shared state (SyspurposeCache)
+      - The current values on the server
+      - The current values on the file system
+    """
+
+    def __init__(self):
+        self.report = SyspurposeSyncActionReport()
+        self.cp_provider = inj.require(inj.CP_PROVIDER)
+        self.uep = self.cp_provider.get_consumer_auth_cp()
+
+    def perform(self):
+        """
+        Perform the action that this Command represents.
+        :return:
+        """
+        self.sync()
```

Comment:
The `self.report` is not updated here despite something was changed in `/etc/rhsm/syspurpose/syspurpose.json` or on candlepin server. We should update following items of `self.report`: `_status, _status` (in case of exceptions: `_exceptions`).